ci/xbuild/extension.py

```diff
@@ -0,0 +1,679 @@
+        """
+        Scan file `src_file` and detect its list of includes.
+
+        The includes are lines of form `#include "header.h"`, where
+        the file `header.h` may be either in the same directory as
+        the source file, or relative to one of the "include"
+        directories.
+
+        This method will resolve the list of includes to their
+        normalized paths, and stores the list into the
+        `self._src_includes` map.
+        """
+        src_path = os.path.dirname(src_file)
+        includes = []
+        with open(src_file, "rt") as inp:
+            for line in inp:
+                mm = re.match(rx_include, line)
+                if mm:
+                    hfile = mm.group(1)
+                    hfile1 = os.path.join(src_path, hfile)
+                    if os.path.exists(hfile1):
+                        includes.append(hfile1)
+                    else:
+                        found = False
+                        for incl_dir in self.compiler.include_dirs:
+                            hfile2 = os.path.join(incl_dir, hfile)
+                            if os.path.exists(hfile2):
+                                includes.append(hfile2)
+                                found = True
+                                break
+                        if not found:
+                            raise ValueError("Cannot locate header file `%s` "
+                                             "included from `%s`"
+                                             % (hfile, src_file))
+        self._src_includes[src_file] = includes
+        self.log.report_src_includes(src_file, includes)
+
+
+    def _compile_files(self):
+        """
+        In this step we compile all the files in the
+        `self._files_to_compile` list.
+
+        Compilation is done in-parallel, using multiple processes
+        (up to `self.nworkers`). All errors/warnings are reported at
+        the end. A SystemExit exception is raised if any of the files
+        fail to compile. The exception, however, is deferred to a
+        point when all files were attempted to be compiled, or at
+        least until a sufficient volume of errors was produced (up to
+        `self.max_error_lines` lines).
+
+        At the end of this step, if it succeeds, all source files will
+        have their corresponding object files physically present on
+        disk.
+        """
+        self.log.step_compile(self._files_to_compile)
+
+        def compile_queue():
+            # Given a queue of workers (`subprocess.Popen` processes),
+            # wait until at least one of them finishes and return that
+            # worker, and also remove it from the `queue`.
+            def await_one(queue):
+                while queue:
+                    for i, worker in enumerate(queue):
+                        if worker.poll() is not None:
+                            del queue[i]
+                            return worker
+                    time.sleep(0.1)
```

Comment:
So it seems that when we have a lot of output from the subprocess, this loop becomes infinite. `worker.poll()` doesn't deadlock, it just never becomes equal to`None`. For example, on Windows I see the following 7 subprocesses that never finish according to `.poll()`:

```
c\sort.cc <subprocess.Popen object at 0x00000210CE6F1F98>
c\column\sentinel_fw.cc <subprocess.Popen object at 0x00000210CE6F1E10>
c\column\sentinel_str.cc <subprocess.Popen object at 0x00000210CE6F1C18>
c\frame\rbind.cc <subprocess.Popen object at 0x00000210CE6F1F28>
c\frame\__sizeof__.cc <subprocess.Popen object at 0x00000210CE6F1D30>
c\models\aggregator.cc <subprocess.Popen object at 0x00000210CE6F1BE0>
c\models\dt_ftrl.cc <subprocess.Popen object at 0x00000210CE6F13C8>
```

In reality these subprocesses did produce the `.obj` files (and I guess stopped).
What is also true is that these files produce a lot of warnings, so the pipe could cause some problems.

Reply:
Yep, that's exactly what I was talking about yesterday -- the pipe fills up and the process dead-locks. However, that's only if a single file emits voluminous amount of warnings, which normally doesn't happen. Still, we should fix this in pr `xbuild2`

Reply:
Yes, I thought you've been using `.wait()` instead of `.communicate()` somewhere, but it seems that even `.poll()` doesn't work as expected in such cases.

Reply:
May be we also add a variable that controls the maximum time allowed for compiling a single file. Even when we solve the pipe problem, the loop can potentially become infinite if compilation hangs at some stage.

Reply:
Never heard of a compiler hanging

Reply:
May not be a direct compiler hang, but if the `while` loop can go infinite, it will at some point :)